managed/services/management/role.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package management
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm/api/managementpb"
+	"github.com/percona/pmm/managed/models"
+)
+
+// ErrInvalidRoleData is returned when a row cannot be asserted to role.
+var ErrInvalidRoleData = fmt.Errorf("InvalidRoleData")
+
+// RoleService represents service for working with roles.
+type RoleService struct {
+	db *reform.DB
+
+	managementpb.UnimplementedRoleServer
+}
+
+// NewRoleService creates a RoleService instance.
+func NewRoleService(db *reform.DB) *RoleService {
+	//nolint:exhaustruct
+	return &RoleService{
+		db: db,
+	}
+}
+
+// CreateRole creates a new Role.
+func (r *RoleService) CreateRole(_ context.Context, req *managementpb.RoleData) (*managementpb.RoleID, error) {
+	var role models.Role
+	role.Title = req.Title
+	role.Filter = req.Filter
+
+	if err := models.CreateRole(r.db.Querier, &role); err != nil {
+		return nil, err
+	}
+
+	return &managementpb.RoleID{
+		RoleId: role.ID,
+	}, nil
+}
+
+// UpdateRole updates a Role.
+//
+//nolint:unparam
+func (r *RoleService) UpdateRole(_ context.Context, req *managementpb.RoleData) (*managementpb.EmptyResponse, error) {
+	err := r.db.InTransaction(func(tx *reform.TX) error {
+		var role models.Role
+		if err := tx.FindByPrimaryKeyTo(&role, req.RoleId); err != nil {
+			if errors.As(err, &reform.ErrNoRows) {
+				return status.Errorf(codes.NotFound, "Role not found")
+			}
+			return err
+		}
```

Comment:
Not sure if the transaction will protect the selected record from a concurrent update (in case that was the original purpose of the transaction), since under the hood it's the SELECT, not the SELECT FOR UPDATE.

Reply:
We could actually get away without any transaction to decrease the overhead.
The `FindByPrimaryKeyTo` is there only to check if the record exists. `Update` operation is atomic.
What do you think?

Reply:
If there is no additional business logic involved (like e.g. having at least one role with a particular title or filter) then I don't see any problems in removing the transaction.